README.txt
DXCONVERT/TXCONVERT version 3.x
date: 20210508

===================================

WHAT'S THIS? 

DXCONVERT
is a cross-platform tool that reads files with DX7(II)/TX7 (compatible) 
patchdata, and writes them to files with DX7/DX7II/TX7 patchdata as pure 
clean SysEx data, standard format 0 MIDI file with SysEx data, or 
4096 bytes packed 32-voice DX7 format (VMEM). DXCONVERT can read many many 
different input file formats, including some strange and exotic ones. 
If in doubt, simply try. If the output is garbage, the fileformat is not 
or badly supported. For a list of currently supported file and data 
formats see FORMATS.txt. 
If you encounter an unsupported DX7 fileformat: please report!

DXconvert can also read files with voicedata from 4-op FM synth
(TX81Z, DX21, FB01, etc.), 2-op FM synths (PSS-480/580/680), and
.opm files from VOPM (a free 4-op FM VST plugin) and converts
these to DX7 voices. 

Optionally DXconvert can perform all kinds of manipulations on the data like:
sort by name, randomize, remove duplicates, find by name or select by number.

TXCONVERT
is a similar tool but it writes data to be used with the Yamaha 4-OP FM synths:
DX100, DX27(s), DX21, DX11 (V2), TX81Z, WT11, DS55, TQ5, YS100, YS200, B200, V50, FB01,
and Reface DX, and with a free VST plugin called VOPM.
It can read the same 4-op and 2-op FM synth voicedata that DXconverts reads,
but not the 6-OP DX7 data. This README file is written for DXconvert. A better
description of TXconvert is on my TODO list, but reading this document should
give you some idea about the use of TXconvert also.

====================================

REQUIREMENTS?

These tools were written using Python 3.8 
If you install the latest stable version of 3.x it should work.
It might also work with Python 2.7, but this is not guaranteed anymore.

Python is available for free for all major platforms ( Windows, Mac, Linux).

For MIDI support python-rtmidi is required. It's available from: 
    http://pypi.python.org/pypi/python-rtmidi/

For import of TX7/DX21/DX100/DX11 cassette-interface data in wav or cas
format you need the MSX castools from Vincent vam Dam. You can find it
here: http://home.kabelfoon.nl/~vincentd/
Also read README-wav2cas.txt

====================================

INSTALLATION? 

- Install Python-3.4.0 (or newer) for your OS
- Make sure python c.q. python.exe is in your system PATH

After installation of Python, you can copy the DXconvert directory to a 
suitable location. The subdirectory named "DXconvert" inside that 
directory should be in the directory from where you start dxconvert.py, 
or it should be a location where python on your system searches for 
packages and modules. (PYTHONPATH). For example on my Linux Fedora 
system, I have copied the DXconvert subdirectory to 
${HOME}/.local/lib/python-3.7/site-packages/. And I have copied/renamed 
dxconvert.py to ~/bin/dxconvert in my HOME directory. 
The ideal situation is when dxconvert.py is in your systems PATH 
and the DXconvert subdirectory in your PYTHONPATH.

If you have installed python-rtmidi MIDI I/O support is available.
Running midi_help.py script will show you the available MII ports.
You can select the ones you need and create an initial dxtxmidi.cfg
file. You can also use the main dx/txconvert programs to select the
MIDI ports. In simple situations this only needs to be done once.

NOTE: I do no not know much about Windows and Mac. Please add information 
to this README file and send it to me if you have used DXconvert 
succesfully on your platform.


=======================================

USAGE? 

GUI usage:
----------
Double-click on dxconvert-gui.py.
Select one or more input files, and one output filename. 

You can select DX7 or DX7II compatibility. If you deselect the DX7II checkbox, 
you might loose some parameters that only the DX7II/TX802/DX200 support.
DX7II data can also be exported as TX7 performance data, and vice versa.

The export fileformat is based on the extension of the output file name. 
4 file extensions are allowed:

1) DXconvert: .dx7 or .DX7:
4096 bytes raw DX7 voice data, without headers, without DX7II 
additional data. Do not use this fileformat if you want to keep DX7II 
additional parameters (AMEM/ACED) If the input file contains only a 
single patch, the patch is saved as a 128 bytes small single VMEM 
datapack. The program tries to convert DX7II/DX200 "AMS" and "PEGR"
parameters into old DX7 AMS and PEG data for improved compatibility. 

TXconvert: use *.dxx for raw 4OP voice data, *.fb1 for raw FB01 data.

2) .syx or .SYX:
DX7, TX7, or DX7II voice data SysEx file including SysEx header, 
checksum, and EOX (End Of Exclusive 0xf7).

3) .mid, .midi, .MID, or .MIDI: 
Same data as SysEx, but written as a Standard Midifile 
format 0 track that can be imported directly into your MIDI sequencer.

4) .txt or .TXT:
- If there is only one selected voice, this will save a list of all 
voiceparameters and their values. (decimal MIDI values, not always 
values as displayed on DX7)
- If there are more selected voices, this will save a list with 
voicenames.

5) other or no extension: SysEx format is used.

6) Special: if "MIDI" (without the quotes, case-sensitive) is selected as 
outfile name, data will be sent to a MIDI port, if available. To select 
the right MIDI port click on the MIDI icon.  

Many additional options are available like sorting, selecting, and
randomizing patches. Just click on the switches or fill in the appropiate
fields. See below for a description of the options (COMMANDLINE usage). 

If the input file contains more than 32 patches (e.g. 64 or 128) the 
output is split into multiple files. They will have names like 
INFILE(1).syx INFILE(2).syx etc. (Unless you select the --nosplit option)

NOTE: After completing your conversion you are free to change the file 
extension from .dx7 or .syx into something else. Some stupid DX7 managers 
refuse to read files if they do not have a specific extension. I have 
seen extensions like .bnk or .32 for simple SysEx or raw DX7 files. 
DXconvert can read all of them, at least all the ones that I have seen.


COMMANDLINE usage (Recommended!):
---------------------------------
The commandline version is named dxconvert.py/txconvert.py, 
The GUI version dxconvert-gui.py/txconvert-gui.py (or .pyw) has exactly the
same functionality.

Usage: 
	dxconvert.py INFILE(S) OUTFILE [options]

Commandline options:	
    dxconvert.py --help
    txconvert.py --help
will tell you many more tricks you can do with dxconvert/txconvert 

===========================================================================

LIMITATIONS and BUGS:

- Only DX7/DX7II Voice data and TX7 performance data are supported. 
Other data like DX7II Performances, Microtunings, and Fractional Scaling 
data are ignored and not saved in the exported files. 

- On some systems (Windows ?) splitting very large libraries is reported
to fail in combination with very long pathnames when using pythonw.exe to
start DXconvert. I have no idea why. Maybe a pythonw.exe bug ?
It does not seem to happen on my Linux system, and it does not happen 
when using python.exe instead of pythonw.exe to start DXconvert on 
Windows. That's why since 1.1.10 I use ".py" instead of ".pyw" as extension
for txconvert-gui.py and dxconvert-gui.py

Note: It seems that with Python-3.x version this problem has been fixed.

ALWAYS KEEP A BACKUP OF YOUR ORIGINAL FILES! 
I did a lot of testing, but still there could be bugs. 
Let me know if you find any. Also report unsupported Yamaha DX/TX
voice data fileformats to me. It might be supported in a next version.


========================================================================


EXTRAS:

In the Tools directory you will some useful utilities like mid2syx.py and 
syx2mid.py You can use these to convert ANY SysEx file into a MIDI format 0
file, or to extract SysEx data from a MIDI file. Use this for example if you 
want to convert a DX7II file including Performance and Microtuning data to 
standard MIDI file. But you can also use it with files from other synths.

=========================================================================

THANKS:

Many thanks go to Charles Copp (DX200 import ideas), Renata Dokmanovic (a 
lot), Sean Bolton (FB01 import and Hexter-DSSI, which I used a lot for 
testing sounds.), Paul Deco (heavy testing), Daniel Forró (exotic file examples),
Jon Morgan (FM-ALive DX Manager), Sam (VOPM), Steve Finney (Reface DX testing), 
ChunkyBloke (Korg DS-8 testing and conversion improvements)
and several members of Yahoo forums and Facebook groups who helped me with facts, 
figures, file examples, bugreports, feature requests, PayPal donations and good 
ideas to create and improve this program.

==========================================================================

LEGAL:

Yamaha is a registered trademark etcetera etcetera blablablabla ...
DXconvert is published under GPL v3 or higher. See LICENCE.txt 

=========================================================================

DONATIONS:

Do you want to support this project? 
Donations are welcome: 
https://paypal.me/mtarenskeen

==========================================================================

CONTACT:

m.tarenskeenATkpnmail.nl

or join the Facebook Yamaha DX7 owners, Yamaha Reface, Yamaha Soundmondo, Korg DS8 & 707, and/or Yamaha V50 group(s), 
or join the YamahaDX forum at groups.io 

Enjoy DXconvert and have more fun with your Yamaha synths!

M.T.



CHANGES.txt
DXconvert HISTORY
-----------------
3.1.6 (20210606)
- Chorus FX low rate, high depth on RefaceDX to match DX21/DX27s Chorus
- minor bugfixes

3.1.5 (20210604)
- Improved LFO conversion based on new measurements
- some bugfixes

3.1.4 (20210515)
- Improvements in 4OP->DX7 EG conversions

3.1.3 (20210507)
3.1.2 (20210424)
- More improvements KORG 707/DS8 conversions

3.1.1 (20210420)
- Some improvements in KORG 707/DS8->DX/TX conversions (Thanks to ChunkyBloke)

3.1.0 (20210316)
- Support for undocumented, but very usable, Reface DX bankdump format. 
- Limit Reface DX PEG Level range -/+ 48

3.0.8 (20201027)
- fixed bug reading JSynthLib patchlib.xml data with Python3
- improved Reface DX TXT parameter list output.

3.0.7 (20200626)
- Reworked TX81Z (etc) to RefaceDX Feedback level conversion
- Reworked TX81Z Operator wave to RefaceDX Feedback approximations
- Reworked V50/YS200 -> Reface DX FX levels
- More work on RefaceDX level scaling 
- Small changes Reface and TX81Z parameterlist txt output

3.0.6
- Reworked TX81Z (etc) to RefaceDX operator output level conversions
- Reworked TX81Z (etc) to RefaceDX level scaling conversion 

3.0.5
- Set FX parameter values to 64 if effect is switched off (initvoice values)
- MIDI options disabled if python-rtmidi is not installed

3.0.4 (20200516)
- Lowest value Reface DX outlevel=0, not 2 in conversion table.
- Fix range mode V50 low range frequency = normal fixrange frequency / 300
- Use formula instead of lookuptables to calculate fixed frequencies

3.0.3 (20200422)
- Carrier output level for TX81Z and RefaceDX have a algorithm-dependent carrierlevel adjustment.
    Bugfix: when converting from TX81Z to RefaceDX no extra adjustment is needed.  
- DX/TX/RefaceDX detune conversion refined
- RefaceDX simplified Level Scaling conversion
- python3 vs python2 character encoding error workaround in vopm.py
- added synthtype "V2". In Japan the DX11 was named V2.
- Small changes in documentation

3.0.0 (20191212)
- Moving on to Python3, Python2.x compatibility not guaranteed.
- six.py no longer required.
- Tools/syx2mid.py new option --trackname
- Tools/mid2syx.py check if inputfile is a midifile first

2.3.7 (20190505)
- TX81Z/DX11 Librarian (Bowker/Henderson/Michelson) .PAT import 

2.3.6 (20181102)

2.3.5 (20180427)
- --fc1, --fc2 and --bc commandline switches now default to True.
- Fixed forgotten activation of --brightness option in dxconvert :-)
- -bc and -bc2at options (known from dxconvert) now also available for txconvert
- Musicode Voice Development System file (.VBL) import
- attempt to fix TQ5_vmem.req

2.3.4 (20180226)
- Time for a subversion bump
- Fixed RefaceDX LFO speed bug
- Fixed minor bug in Korg DS8/707 LFO speed conversion 
- Refresh random.seed() before randomizing
- TFM instrument file (.tfi) import
- added some tooltip strings
- fixed sort/sort2 buttons in dxconvert-gui

2.3.3 (20171118)
- Tools/v50convert.py for conversion to/from V50 Floppy data.
- convert some more CX5M cas files.
- Big Bug fix: --nosilence option would kill all good instead of all silent sounds :-) 
- Improved OPM file parser to read also Luke McQueen's collection
- New option --exclude. Usage same as --find, but inverted effect.
  commandline only.
- Improved identification of Reface DX sysex data.
- Fixed Korg DS8/707 single voice import.
- Fixed txt (parameter listing) output when input is single patch
- "On/Off" value for DX21 Chorus in parameter list
- Fix bug with --split in txconvert-gui.py
- Small changes
- Working on BOHM/ORLA DSE24 import (unfinished, experimental)
- Fixed a few out-of-range crashes
- Small bugfixes in PSS480 conversion

2.3.3 (20170221)
- remove tkinter requirement for commandline versions
- import six library only for Tkinter gui 
- gui start without rtmidi fixed

2.3.2 (20161128)
- bump version to 2.3.2
- fix bug with OPM export from txconvert-gui
- fixed a bug with --tx7 and --dx72 buttons in dxconvert-gui
- fixed a division-by-zero error (refacedx)
- prevent out-of-range values (refacedx)
- if --outfile is null device, --nosplit=True automatically 
- conversion of fx delay times (Reface, DS-8, V50) more accurate 
- improved conversion of effect level (V50, YS200) to Reface DX
- improved operator outputlevel conversions to/from Reface DX.
- initial experimental support for conversion *from* RefaceDX to others. 
    (experimental, unfinished, and only partially possible.) 
- --nosilence option: also check sustain levels and Operator on/off switches.

2.3.1 (20160820)
- New option --nosilence to remove patches that produce no sound
- GUI improvements:
    - --sort2 (case-sensitive sorting) option added to GUI.
    - --yamaha option (txconvert) now uses a dropdown list.
    - Tooltip help 
    - new MIDI icon (only visible if python-rtmidi is installed)
- project name now simply DXconvert again, instead of DXconvert-six
- more types of Galaxy TX81Z files supported
- Midiquest TX81Z .SQL
- stricter check for valid filenames before export
- Tolerate Korg Volca FM non-standard 164 bytes DX7 VCED SysEx 
- optimize conversions from DX9 EG to AR-D1R-D1L-D2R-RR. 
- improved EG/rate scaling conversion for Reface (still not perfect)
- improved LFO delay range conversion approximation (todo: more info/measurements)
- removed ugly TIMBRE_ADJUST and FB_ADJUST variables and code from fourop.py.
  (Not needed anymore, because the --brightness option is much more elegant)
- fixed Reface portamento mode
- TX81Z and PSS waveform simulation using Reface Feedback (experimental)
- Fixed reversed EG levels in pss480 conversion
- Fixed algorithm #5 bug in fourop.carrier()
- Convert DX21 chorus to Reface DX chorus effect.
- Lots of minor code cleanup changes
- Some bugfixes in dxconvert-gui and txconvert-gui
- Some minor bugfixes

2.3.0beta2 (20160421)
- Metrasound DX Supervisor import

txconvert.py
#!/usr/bin/env python3

"""
txconvert.py

(c)2020 Martin Tarenskeen <m.tarenskeenATkpnmailDOTnl>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""

__license__ = 'GPL v3'

import sys
import os
from argparse import ArgumentParser
from glob import glob
from DXconvert import TXC
from DXconvert import dxcommon
try:
    import rtmidi
    ENABLE_MIDI = True
    del rtmidi
except:
    ENABLE_MIDI = False

PROGRAMNAME = TXC.PROGRAMNAME
PROGRAMVERSION = dxcommon.PROGRAMVERSION
PROGRAMDATE = dxcommon.PROGRAMDATE

############# CLI ###################

def cli_main(argv=sys.argv):
    progname = os.path.basename(argv[0])

    parser = ArgumentParser(
            description = '{}\nVersion: {} ({})'.format(PROGRAMNAME, PROGRAMVERSION, PROGRAMDATE), )
    parser.add_argument('-b', '--brightness', default=0, type=int, help='Adjust global brightness (+/-)')  
    parser.add_argument('-bc', '--breathcontrol', action='store_false', default=True, help='Do NOT use BreathControl')
    parser.add_argument('-bc2at', '--bc2at', action='store_true', default=False, help='Copy BreathControl data to Aftertouch')
    parser.add_argument('-c', '--channel', default=0, type=int, help='Midi channel (1~16)in SysEx header')
    parser.add_argument('-C', '--check', action='store_true', default=False, help='Check SysEx checksum before import') 
    parser.add_argument('-f', '--find', metavar='STRING', help='Search for STRING in patchnames')
    if ENABLE_MIDI:
        parser.add_argument('-mi', '--mid_in', help='select midiport MID_IN to receive data FROM synth when selecting a .req file')
        parser.add_argument('-mo', '--mid_out', help='select midiport MID_OUT to send data TO synth when choosing "MIDI" as outfile')
        parser.add_argument('-m', '--mid', help='use this option as a shortcut for "--mid_in MID_IN --mid_out MID_OUT", if MID_IN and MID_OUT have the same name MID')
    parser.add_argument('-n', '--nosplit', action='store_true', default=False, help="Don't split: save data in one file")
    parser.add_argument('-nd', '--nodupes', action='store_true', default=False, help='Remove duplicates')
    parser.add_argument('-nd2', '--nodupes2', action='store_true', default=False, help='Remove duplicates, also with different names')
    parser.add_argument('-ns', '--nosilence', action='store_true', default=False, help='Remove patches that produce no sound')
    parser.add_argument('-o', '--offset', default='0', help="Ignore first/last (+/-) OFFSET bytes in input file(s)")
    parser.add_argument('-r', '--random', action='store_true', default=False, help="easy randomizer") 
    parser.add_argument('-sp', '--split', action='store_true', default=False, help="Save each patch in a separate file")
    parser.add_argument('-s', '--select', metavar='RANGE', help='Select patch RANGE to save')
    parser.add_argument('-S', '--sort', action='store_true', default=False, help='Sort patches by name, not case-sensitive')
    parser.add_argument('-S2', '--sort2', action='store_true', default=False, help='Sort patches by name, case-sensitive')
    parser.add_argument('--copy', nargs=2, metavar=('X', 'Y'), help='Copy patch nr X to nr Y')
    parser.add_argument('--swap', nargs=2, metavar=('X', 'Y'), help='Swap patch nr X and nr Y')
    parser.add_argument('-v', '--version', action='version', version='{} ({})'.format(PROGRAMVERSION, PROGRAMDATE)) 
    parser.add_argument('-x', '--exclude', metavar='STRING', help='Exclude if STRING is found in patchname')
    parser.add_argument('-y', '--yamaha', metavar='MODEL', default='tx81z', help='Specify Yamaha synth MODEL for outfile') 
    parser.add_argument('infile', nargs='+', help='Selects input file(s)')
    parser.add_argument('outfile', help='Select output file')

    args=parser.parse_args()
    infilez = args.infile
    infiles = []
    for i in infilez:
        infiles += glob(i)

    outfile = args.outfile
    outfile_ext = os.path.splitext(outfile)[1]
    outfile_dir = os.path.split(outfile)[0]

    mid_in = ""
    mid_out = ""
    if ENABLE_MIDI:
        mid_in = dxcommon.MID_IN
        mid_out = dxcommon.MID_OUT
        CFG = dxcommon.CFG
        if os.getenv('MID_IN'):
            mid_in = os.getenv('MID_IN')
        mid_out = os.getenv('MID_OUT')
        if os.path.exists(CFG):
            with open(CFG, 'r') as f:
                for line in f.readlines():
                    l = line.split('=')
                    if l[0].strip() == 'MID_IN':
                        mid_in = l[1].strip()
                    if l[0].strip() == 'MID_OUT':
                        mid_out = l[1].strip()

        if args.mid:
            mid_in = args.mid
            mid_out =  args.mid
        if args.mid_in:
            mid_in = args.mid_in
        if args.mid_out:
            mid_out = args.mid_out
        if args.mid_in or args.mid_out:
            with open(CFG, 'w') as f:
                if mid_in: f.write('MID_IN = {}\n'.format(mid_in))
                if mid_out: f.write('MID_OUT = {}\n'.format(mid_out))

    bc = args.breathcontrol
    bc2at = args.bc2at
    brightness = args.brightness
    nosplit = args.nosplit
    ch = args.channel
    check = args.check
    select = args.select
    sort = args.sort
    sort2 = args.sort2
    swap = args.swap
    copy = args.copy
    nodupes = args.nodupes
    nodupes2 = args.nodupes2
    nosilence = args.nosilence
    offset = args.offset
    Random = args.random
    find = args.find
    exclude = args.exclude
    yamaha = args.yamaha.lower().strip()

    FB01 = False
    REFACE = False
    if yamaha in ('ys100', 'ys200', 'tq5', 'b200', 'ds55', 'v50'):
        split = 25
    elif yamaha in ('dx27', 'dx27s', 'dx100'):
        split = 24
    elif yamaha == 'fb01' or outfile_ext.lower() == '.fb1':
        yamaha = 'fb01'
        FB01 = True
        split = 48
    elif yamaha in ('opm', 'vopm') or outfile_ext.lower() == '.opm':
        yamaha = 'vopm'
        FB01 = True
        split = 128
    elif yamaha in ('reface', 'dxreface', 'refacedx'):
        yamaha = 'refacedx'
        REFACE = True
        split = 32
    else:
        split = 32
    
    if args.split:
        split = 1

    if nosplit or (outfile == "MIDI") or (outfile == os.devnull):
        split = sys.maxsize
        nosplit = True

    txdata = []
    for infile in infiles:
        print ("Reading {}".format(infile))
        if not os.path.exists(infile):
            print ("{} not found".format(infile))
            return 1
        if infile == outfile:
            print ("Must have different input and output files")
            return 1

        if os.path.isfile(infile):
            txdat, channel = TXC.read(infile, offset, check, yamaha, mid_in, mid_out)
            txdata += txdat

    if ch != None:
        channel = ch-1
        channel = max(0, channel)
        channel = min(15, channel)

    if nodupes2:
        nodupes = True
            
    if nodupes or nodupes2:
        if FB01:
            txdata = TXC.fbnodupes(txdata, nodupes2)
        elif REFACE:
            txdata = TXC.rdxnodupes(txdata, nodupes2)
        else:
            txdata = TXC.txnodupes(txdata, nodupes2)

    if nosilence:
        if FB01:
            txdata = TXC.fbnosilence(txdata)
        elif REFACE:
            txdata = TXC.rdxnosilence(txdata)
        else:
            txdata = TXC.txnosilence(txdata)

    if select != None: 
        txdat = []
        for i in dxcommon.range2list(select):
            if FB01:
                txdat += txdata[64*(i-1): 64*i]
            elif REFACE:
                txdat += txdata[150*(i-1): 150*i]
            else:
                txdat += txdata[128*(i-1): 128*i]
        txdata = txdat
    
    if args.find != None:
        if FB01:
            txdata = TXC.fbfind(txdata, find)
        elif REFACE:
            txdata = TXC.rdxfind(txdata, find)
        else:
            txdata = TXC.txfind(txdata, find)
    
    if args.exclude:
        if FB01:
            txdata = TXC.fbexclude(txdata, exclude)
        elif REFACE:
            txdata = TXC.rdxexclude(txdata, exclude)
        else:
            txdata = TXC.txexclude(txdata, exclude)
        
    if brightness:
        if FB01:
            txdata = TXC.fbbrightness(txdata, brightness)
        elif REFACE:
            txdata = TXC.rdxbrightness(txdata, brightness)
        else:
            txdata = TXC.txbrightness(txdata, brightness)

    if bc2at:
        if (FB01 == False) and (REFACE == False):
            for i in range(len(txdata)//128):
                txdata[128*i+84:128*i+88] = txdata[128*i+53:128*i+57]
                if txdata[128*i+86]>50:
                    txdata[128*i+86] -= 51
                else:
                    txdata[128*i+86] +=50

    if bc == False:
        if (FB01 == False) and (REFACE == False):
            for i in range(len(txdata)//128):
                txdata[128*i+53:128*i+57] = [0,0,50,0]
    
    if Random:
        if FB01:
            txdata = dxcommon.dxrandom(txdata, 64)
        elif REFACE:
            txdata = dxcommon.dxrandom(txdata, 150)
        else:
            txdata = dxcommon.dxrandom(txdata, 128)

    if sort:
        if FB01:
            txdata = TXC.fbsort(txdata, False)
        elif REFACE:
            txdata = TXC.rdxsort(txdata, False)
        else:
            txdata = TXC.txsort(txdata, False)

    if sort2:
        if FB01:
            txdata = TXC.fbsort(txdata, True)
        elif REFACE:
            txdata = TXC.rdxsort(txdata, True)
        else:
            txdata = TXC.txsort(txdata, False)

    if copy:
        if FB01:
            txdata = TXC.txcopy(dxcommon.range2list(copy[0]), int(copy[1]), txdata, 64)
        elif REFACE:
            txdata = TXC.txcopy(dxcommon.range2list(copy[0]), int(copy[1]), txdata, 150)
        else:
            txdata = TXC.txcopy(dxcommon.range2list(copy[0]), int(copy[1]), txdata, 128)

    if swap:
        if FB01:
            txdata = TXC.txswap(int(swap[0]), int(swap[1]), txdata, 64)
        elif REFACE:
            txdata = TXC.txswap(int(swap[0]), int(swap[1]), txdata, 150)
        else:
            txdata = TXC.txswap(int(swap[0]), int(swap[1]), txdata, 128)

    if FB01:
        n = len(txdata)//64
    elif REFACE:
        n = len(txdata)//150
    else:
        n = len(txdata)//128
        
    if (split == 1) and (n > 1):            
        if FB01:
            for i in range(n):
                outfile_name = dxcommon.list2string(txdata[64*i:64*i+7])
                outfile_name = dxcommon.validfname(outfile_name)
                Outfile = os.path.join(outfile_dir, outfile_name + outfile_ext)
                
                count = 0
                while os.path.exists(Outfile):
                    count += 1
                    if count>1:
                        Outfile = os.path.join(outfile_dir, outfile_name + "(" + str(count) + ")" + outfile_ext)

                print (TXC.write(Outfile, txdata[64*i:64*(i+1)], channel, nosplit, 1, yamaha, mid_out))
        elif REFACE:
            for i in range(n):
                outfile_name = dxcommon.list2string(txdata[150*i:150*i+10])
                outfile_name = dxcommon.validfname(outfile_name)
                Outfile = os.path.join(outfile_dir, outfile_name + outfile_ext)
                
                count = 0
                while os.path.exists(Outfile):
                    count += 1
                    if count>1:
                        Outfile = os.path.join(outfile_dir, outfile_name + "(" + str(count) + ")" + outfile_ext)
                
                print (TXC.write(Outfile, txdata[150*i:150*(i+1)], channel, nosplit, 1, yamaha, mid_out))
        else:
            for i in range(n):
                outfile_name = dxcommon.list2string(txdata[128*i+57:128*i+67])
                outfile_name = dxcommon.validfname(outfile_name)
                Outfile = os.path.join(outfile_dir, outfile_name + outfile_ext)
                
                count = 0
                while os.path.exists(Outfile):
                    count += 1
                    if count>1:
                        Outfile = os.path.join(outfile_dir, outfile_name + "(" + str(count) + ")" + outfile_ext)
                
                print (TXC.write(Outfile, txdata[128*i:128*(i+1)], channel, nosplit, 1, yamaha, mid_out))
    else:
        print (TXC.write(outfile, txdata, channel, nosplit, split, yamaha, mid_out))

    return 0

################### MAIN #####################

if __name__ == '__main__':
    sys.exit(cli_main())
    

